### ISSUE TITLE
Offer to textart.

### ISSUE THREAD

**Issue by clort81** [razterizer/Termin8or#43]
Hope to someday contribute to terminal based game art.
![realmsOAsamplecap](https://github.com/user-attachments/assets/b1927c81-aa90-4fd5-b0d6-78065545a485)
Just issuing ECMA-48 directly myself, since all i need is clear screen  and gotoxy.
Not getting very far on my own.

**Comment by razterizer** [comment]
Hi clort81!
I'm very happy for your engagement with my API. That artwork of yours look really nice!!
I presume this is pixel art using 8-bit colors? Am I correct? I see you use the ANSI file format. I'm not able to find much info on this format so I'm curious on how those pixels are produced. Is it by tiling certain characters from different Windows code pages or how? I found this page: https://en.wikipedia.org/wiki/ANSI_character_set telling me "ANSI character set" can refer to either Windows-1252 or Code Page 437. As I cannot find much info about the ANSI file format (yet), it is hard to try to imagine how to extend my API to work with it. Also as my API is cross-platform (Linux/Mac/Win), this can probably lead to compatibility issues.

`Termin8or` currently uses the standard ASCII character set with 4-bit colors, so there quite a gap needed to be bridged to be able to do this unfortunately.

I have thought about adding support for 8-bit colors (Windows Console has support for it after Windows 10 build 14931), so that's in my mental TODO-list for now. 

I have also thought about adding support for extended characters such as extended ASCII, UTF-8 or some semi-custom(?) mapping similar to that of `RexPaint`. I'm not sure how to do design this properly yet though.

Merry Christmas!

**Comment by razterizer** [comment]
256 colors is coming very soon. Almost finished implementation.

After that, I plan to add utf-8 support.

**Comment by clort81** [comment]
Happy you made this. +✓ and <3

**Comment by razterizer** [comment]
8-bit colors have been done already some time ago (259 including two transparency modes and default colors) and now I'm happy to announce that unicode support is more or less done, but before I merge I first have to finish the GlyphPicker I currently work on and after that I plan to do ANSI-art import to Texture data structure and vice versa. If possible maybe even RexPaint import, but I expect that to be a bit complex. 

Also, I am beginning to realize that text art may be dependent on what font is used - e.g. by texture mapping, like the polyduck text art fonts. This could be a bit tricky to implement, because then (I'm guessing) there could be a discrepancy between what unicode character the programmer intended and what glyph is actually mapped out by the codepoint via the font.

Anyways, the way the system works on branch `feature/terminator-utf8` is that you can use the old text rendering scheme as normal if using `ScreenHandler<char>` (just as optimal as before) or you can use the new unicode supported renderer by `ScreenHandler<char32_t>` and the neat thing is that you can choose if you want to input `char`, `std::string`, `t8::Glyph` or `t8::GlyphString` and no matter if you use the old or new renderer it just works. `t8::Glyph` contains a unicode codepoint and a `char` for ASCII fallback. Fallbacks happen if you use the old renderer or if the terminal and/or font used does not support those codepoints that are fed to the renderer.

Just a glance over the current state. 

Hope this is to your liking.
After all, this project started out as a way for me to learn how to use escape codes and grew into a renderer or game engine lib that  seems to bridge the gap between artists like you and game developers. At least that is what I think this has become.

God bless.

**Comment by clort81** [comment]
Not all fonts provide usable glyphs for textart.  It's a herculean effort to update them, so that's a project of mine.

Will get a wide variety of terminal fonts with proper utf-8 block-drawing characters and block-braille.

**Comment by razterizer** [comment]
That is true. Though the way Termin8or "<char32_t> works is by checking for valid glyph cell size and then checking for different fonts if the glyph exists. 

It was tedious to map out, but it really helped with my example program in the Core lib which lists every block in the unicode standard and allows me to filter by block name and glyph width if available.

Termin8or on the utf8 branch now supports the most common fonts on cmd (conhost-like) and with all fonts on MacOS (MacOS seems to combine glyphs if code point is missing for a certain font, so you just have to do the mapping once there). On windows terminal + linux I simply assume decent coverage (YOYO).

So in most cases for good fonts or fairly standard fonts for conhost-like (consolas etc) you automatically get the right fallback ascii character. Note that each fallback ascii cp is custom specified by the API user.

For custom fonts such as polyducks fonts, I currently have no good solution for that. Perhaps a code point/glyph coverage meta file that accompanies the font could be a solution to that.

Hope this clarifies some things.

**Comment by razterizer** [comment]
I've been thinking about your perspective some more and I think I understand your point better now; for text art, fallback characters aren’t really acceptable since they change the visual result. Right?

What I’ve built so far guarantees that something is rendered, but not necessarily that it’s identical across fonts. That’s a slightly different goal than what you might be after.

It might make sense for me to add a kind of "strict mode" where missing glyphs are detected instead of silently falling back, so the artist can find out if the output won’t match the intent on a given setup.

Your idea of improving font coverage and having proper block/braille support fits very well with that direction.

I'll see what can be done about this after I'm finished with the unicode branch.
